--- LLVM/libcxx/test/support/test_iterators.h
+++ oneDPL/test/support/test_iterators.h
@@ -0,0 +1 @@
+// -*- C++ -*-
@@ -1,0 +3,7 @@
+//
+// Copyright (C) Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// This file incorporates work covered by the following copyright and permission
+// notice:
@@ -5 +12,0 @@
-// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
@@ -62,3 +68,0 @@
-
-    template <class T>
-    friend void operator,(const T&, const cpp17_output_iterator&) = delete;
@@ -115,3 +118,0 @@
-
-    template <class T>
-    friend void operator,(const T&, const cpp17_input_iterator&) = delete;
@@ -166,3 +166,0 @@
-
-    template <class T>
-    friend void operator,(const T&, const forward_iterator&) = delete;
@@ -215,3 +212,0 @@
-
-    template <class T>
-    friend void operator,(const T&, const bidirectional_iterator&) = delete;
@@ -276,3 +270,0 @@
-
-    template <class T>
-    friend void operator,(const T&, const random_access_iterator&) = delete;
@@ -283,132 +274,0 @@
-#endif
-
-// Since C++20, a container iterator type that is random access is also required to support three-way comparison.
-// See C++20 [tab:container.req], C++23 [container.reqmts]/39 - /41.
-template <class It>
-class three_way_random_access_iterator {
-  It it_;
-  support::double_move_tracker tracker_;
-
-  template <class U>
-  friend class three_way_random_access_iterator;
-
-public:
-  typedef std::random_access_iterator_tag iterator_category;
-  typedef typename std::iterator_traits<It>::value_type value_type;
-  typedef typename std::iterator_traits<It>::difference_type difference_type;
-  typedef It pointer;
-  typedef typename std::iterator_traits<It>::reference reference;
-
-  TEST_CONSTEXPR three_way_random_access_iterator() : it_() {}
-  TEST_CONSTEXPR explicit three_way_random_access_iterator(It it) : it_(it) {}
-
-  template <class U>
-  TEST_CONSTEXPR three_way_random_access_iterator(const three_way_random_access_iterator<U>& u)
-      : it_(u.it_), tracker_(u.tracker_) {}
-
-  template <class U, class = typename std::enable_if<std::is_default_constructible<U>::value>::type>
-  TEST_CONSTEXPR_CXX14 three_way_random_access_iterator(three_way_random_access_iterator<U>&& u)
-      : it_(std::move(u.it_)), tracker_(std::move(u.tracker_)) {
-    u.it_ = U();
-  }
-
-  TEST_CONSTEXPR_CXX14 reference operator*() const { return *it_; }
-  TEST_CONSTEXPR_CXX14 reference operator[](difference_type n) const { return it_[n]; }
-
-  TEST_CONSTEXPR_CXX14 three_way_random_access_iterator& operator++() {
-    ++it_;
-    return *this;
-  }
-  TEST_CONSTEXPR_CXX14 three_way_random_access_iterator& operator--() {
-    --it_;
-    return *this;
-  }
-  TEST_CONSTEXPR_CXX14 three_way_random_access_iterator operator++(int) {
-    return three_way_random_access_iterator(it_++);
-  }
-  TEST_CONSTEXPR_CXX14 three_way_random_access_iterator operator--(int) {
-    return three_way_random_access_iterator(it_--);
-  }
-
-  TEST_CONSTEXPR_CXX14 three_way_random_access_iterator& operator+=(difference_type n) {
-    it_ += n;
-    return *this;
-  }
-  TEST_CONSTEXPR_CXX14 three_way_random_access_iterator& operator-=(difference_type n) {
-    it_ -= n;
-    return *this;
-  }
-  friend TEST_CONSTEXPR_CXX14 three_way_random_access_iterator
-  operator+(three_way_random_access_iterator x, difference_type n) {
-    x += n;
-    return x;
-  }
-  friend TEST_CONSTEXPR_CXX14 three_way_random_access_iterator
-  operator+(difference_type n, three_way_random_access_iterator x) {
-    x += n;
-    return x;
-  }
-  friend TEST_CONSTEXPR_CXX14 three_way_random_access_iterator
-  operator-(three_way_random_access_iterator x, difference_type n) {
-    x -= n;
-    return x;
-  }
-  friend TEST_CONSTEXPR difference_type
-  operator-(three_way_random_access_iterator x, three_way_random_access_iterator y) {
-    return x.it_ - y.it_;
-  }
-
-  friend TEST_CONSTEXPR bool
-  operator==(const three_way_random_access_iterator& x, const three_way_random_access_iterator& y) {
-    return x.it_ == y.it_;
-  }
-#if TEST_STD_VER < 20
-  friend TEST_CONSTEXPR bool
-  operator!=(const three_way_random_access_iterator& x, const three_way_random_access_iterator& y) {
-    return x.it_ != y.it_;
-  }
-#endif
-  friend TEST_CONSTEXPR bool
-  operator<(const three_way_random_access_iterator& x, const three_way_random_access_iterator& y) {
-    return x.it_ < y.it_;
-  }
-  friend TEST_CONSTEXPR bool
-  operator<=(const three_way_random_access_iterator& x, const three_way_random_access_iterator& y) {
-    return x.it_ <= y.it_;
-  }
-  friend TEST_CONSTEXPR bool
-  operator>(const three_way_random_access_iterator& x, const three_way_random_access_iterator& y) {
-    return x.it_ > y.it_;
-  }
-  friend TEST_CONSTEXPR bool
-  operator>=(const three_way_random_access_iterator& x, const three_way_random_access_iterator& y) {
-    return x.it_ >= y.it_;
-  }
-#if TEST_STD_VER >= 20
-  friend constexpr std::strong_ordering
-  operator<=>(const three_way_random_access_iterator& x, const three_way_random_access_iterator& y) {
-    if constexpr (std::three_way_comparable<It>) {
-      return x.it_ <=> y.it_;
-    } else {
-      if (x.it_ < y.it_) {
-        return std::strong_ordering::less;
-      } else if (y.it_ < x.it_) {
-        return std::strong_ordering::greater;
-      } else {
-        return std::strong_ordering::equal;
-      }
-    }
-  }
-#endif
-
-  friend TEST_CONSTEXPR It base(const three_way_random_access_iterator& i) { return i.it_; }
-
-  template <class T>
-  void operator,(T const&) = delete;
-
-  template <class T>
-  friend void operator,(const T&, const three_way_random_access_iterator&) = delete;
-};
-#if TEST_STD_VER > 14
-template <class It>
-three_way_random_access_iterator(It) -> three_way_random_access_iterator<It>;
@@ -506,3 +365,0 @@
-
-  template <class T>
-  friend void operator,(const T&, const cpp20_random_access_iterator&) = delete;
@@ -602,3 +458,0 @@
-
-  template <class T>
-  friend void operator,(const T&, const contiguous_iterator&) = delete;
@@ -662,3 +515,0 @@
-
-    template <class T>
-    friend void operator,(const T&, const three_way_contiguous_iterator&) = delete;
@@ -776,4 +627 @@
-    template <class T2>
-    friend void operator,(const T2&, const ThrowingIterator&) = delete;
-
-  private:
+private:
@@ -850,4 +698 @@
-    template <class T2>
-    friend void operator,(const T2&, const NonThrowingIterator&) = delete;
-
-  private:
+private:
@@ -883,3 +727,0 @@
-
-    template <class T>
-    friend void operator,(const T&, const cpp20_input_iterator&) = delete;
@@ -923,3 +764,0 @@
-
-  template <class T>
-  friend void operator,(const T&, const cpp20_output_iterator&) = delete;
@@ -1119,14 +958,11 @@
-    template <class T>
-    friend void operator,(const T&, const operation_counting_iterator&) = delete;
-
-  private:
-    constexpr void moved_by(difference_type n) {
-      if (counts_ == nullptr)
-        return;
-      if (n > 0)
-        ++counts_->increments;
-      else if (n < 0)
-        ++counts_->decrements;
-      else
-        ++counts_->zero_moves;
-    }
+private:
+  constexpr void moved_by(difference_type n) {
+    if (counts_ == nullptr)
+      return;
+    if (n > 0)
+      ++counts_->increments;
+    else if (n < 0)
+      ++counts_->decrements;
+    else
+      ++counts_->zero_moves;
+  }
@@ -1351,17 +1186,0 @@
-// The ProxyDiffTBase allows us to conditionally specify Proxy<T>::difference_type
-// which we need in certain situations. For example when we want
-// std::weakly_incrementable<Proxy<T>> to be true.
-template <class T>
-struct ProxyDiffTBase {
-  // Add default `operator<=>` so that the derived type, Proxy, can also use the default `operator<=>`
-  friend constexpr auto operator<=>(const ProxyDiffTBase&, const ProxyDiffTBase&) = default;
-};
-
-template <class T>
-  requires requires { std::iter_difference_t<T>{}; }
-struct ProxyDiffTBase<T> {
-  using difference_type = std::iter_difference_t<T>;
-  // Add default `operator<=>` so that the derived type, Proxy, can also use the default `operator<=>`
-  friend constexpr auto operator<=>(const ProxyDiffTBase&, const ProxyDiffTBase&) = default;
-};
-
@@ -1378 +1196,0 @@
-
@@ -1389 +1207 @@
-struct Proxy : ProxyDiffTBase<T> {
+struct Proxy {
@@ -1399,4 +1216,0 @@
-
-  // Explicitly declare the copy constructor as defaulted to avoid deprecation of the implicitly declared one
-  // because of the user-provided copy assignment operator.
-  Proxy(const Proxy&) = default;
@@ -1440 +1254 @@
-    requires(std::equality_comparable<T> && !std::is_reference_v<T>)
+    requires (std::equality_comparable<T> && !std::is_reference_v<T>)
@@ -1452 +1266 @@
-    requires(std::three_way_comparable<T> && !std::is_reference_v<T>)
+    requires (std::three_way_comparable<T> && !std::is_reference_v<T>)
@@ -1461,16 +1274,0 @@
-  }
-
-  // Needed to allow certain types to be weakly_incrementable
-  constexpr Proxy& operator++()
-    requires(std::weakly_incrementable<std::remove_reference_t<T>>)
-  {
-    ++data;
-    return *this;
-  }
-
-  constexpr Proxy operator++(int)
-    requires(std::incrementable<std::remove_reference_t<T>>)
-  {
-    Proxy tmp = *this;
-    operator++();
-    return tmp;