--- LLVM/libcxx/test/support/test_macros.h
+++ oneDPL/test/support/test_macros.h
@@ -3,0 +4,7 @@
+// Copyright (C) Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// This file incorporates work covered by the following copyright and permission
+// notice:
+//
@@ -6 +12,0 @@
-// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
@@ -10,3 +16,6 @@
-#ifndef SUPPORT_TEST_MACROS_HPP
-#define SUPPORT_TEST_MACROS_HPP
-
+#ifndef _SUPPORT_TEST_MACROS_H
+#define _SUPPORT_TEST_MACROS_H
+
+// Attempt to get STL specific macros like _LIBCPP_VERSION using the most
+// minimal header possible. If we're testing libc++, we should use `<__config>`.
+// If <__config> isn't available, fall back to <ciso646>.
@@ -13,0 +23,6 @@
+# if __has_include("<__config>")
+#   include <__config>
+#   define TEST_IMP_INCLUDED_HEADER
+# endif
+#endif
+#ifndef TEST_IMP_INCLUDED_HEADER
@@ -15 +30 @@
-#    include <version>
+#   include <version>
@@ -17 +32 @@
-#    include <ciso646>
+#   include <ciso646>
@@ -19,6 +34,4 @@
-#else
-#  include <ciso646>
-#endif
-
-#define TEST_STRINGIZE_IMPL(...) #__VA_ARGS__
-#define TEST_STRINGIZE(...) TEST_STRINGIZE_IMPL(__VA_ARGS__)
+#endif
+
+#define TEST_STRINGIZE_IMPL(x) #x
+#define TEST_STRINGIZE(x) TEST_STRINGIZE_IMPL(x)
@@ -78,2 +91 @@
-// Given AppleClang XX.Y.Z, TEST_APPLE_CLANG_VER is XXYZ (e.g. AppleClang 14.0.3 => 1403)
-#define TEST_APPLE_CLANG_VER (__apple_build_version__ / 10000)
+#define TEST_APPLE_CLANG_VER (__clang_major__ * 100) + __clang_minor__
@@ -89,16 +101,11 @@
-#if  __cplusplus <= 199711L
-# define TEST_STD_VER 3
-#elif __cplusplus <= 201103L
-# define TEST_STD_VER 11
-#elif __cplusplus <= 201402L
-# define TEST_STD_VER 14
-#elif __cplusplus <= 201703L
-# define TEST_STD_VER 17
-#elif __cplusplus <= 202002L
-# define TEST_STD_VER 20
-#elif __cplusplus <= 202302L
-# define TEST_STD_VER 23
-#else
-# define TEST_STD_VER 99    // greater than current standard
-// This is deliberately different than _LIBCPP_STD_VER to discourage matching them up.
-#endif
+#    if __cplusplus >= 202002L || _MSVC_LANG >= 202002L
+#        define TEST_STD_VER 20
+#    elif __cplusplus >= 201703L || _MSVC_LANG >= 201703L
+#        define TEST_STD_VER 17
+#    elif __cplusplus >= 201402L || _MSVC_LANG >= 201402L
+#        define TEST_STD_VER 14
+#    elif __cplusplus >= 201103L || _MSVC_LANG >= 201103L
+#        define TEST_STD_VER 11
+#    else // assume C++03
+#        define TEST_STD_VER 3
+#    endif
@@ -117,26 +123,0 @@
-#if TEST_STD_VER >= 11
-# define TEST_ALIGNOF(...) alignof(__VA_ARGS__)
-# define TEST_ALIGNAS(...) alignas(__VA_ARGS__)
-# define TEST_CONSTEXPR constexpr
-# define TEST_NOEXCEPT noexcept
-# define TEST_NOEXCEPT_FALSE noexcept(false)
-# define TEST_NOEXCEPT_COND(...) noexcept(__VA_ARGS__)
-#else
-#   if defined(TEST_COMPILER_CLANG)
-#    define TEST_ALIGNOF(...) _Alignof(__VA_ARGS__)
-#   else
-#    define TEST_ALIGNOF(...) __alignof(__VA_ARGS__)
-#   endif
-# define TEST_ALIGNAS(...) __attribute__((__aligned__(__VA_ARGS__)))
-# define TEST_CONSTEXPR
-# define TEST_NOEXCEPT throw()
-# define TEST_NOEXCEPT_FALSE
-# define TEST_NOEXCEPT_COND(...)
-#endif
-
-#if TEST_STD_VER >= 11
-# define TEST_THROW_SPEC(...)
-#else
-# define TEST_THROW_SPEC(...) throw(__VA_ARGS__)
-#endif
-
@@ -152,30 +132,0 @@
-#  define TEST_STD_AT_LEAST_20_OR_RUNTIME_EVALUATED true
-#else
-#  define TEST_STD_AT_LEAST_20_OR_RUNTIME_EVALUATED (!TEST_IS_CONSTANT_EVALUATED)
-#endif
-
-#if TEST_STD_VER >= 23
-#  define TEST_STD_AT_LEAST_23_OR_RUNTIME_EVALUATED true
-#else
-#  define TEST_STD_AT_LEAST_23_OR_RUNTIME_EVALUATED (!TEST_IS_CONSTANT_EVALUATED)
-#endif
-
-#if TEST_STD_VER >= 26
-#  define TEST_STD_AT_LEAST_26_OR_RUNTIME_EVALUATED true
-#else
-#  define TEST_STD_AT_LEAST_26_OR_RUNTIME_EVALUATED (!TEST_IS_CONSTANT_EVALUATED)
-#endif
-
-#if TEST_STD_VER >= 14
-# define TEST_CONSTEXPR_CXX14 constexpr
-#else
-# define TEST_CONSTEXPR_CXX14
-#endif
-
-#if TEST_STD_VER >= 17
-# define TEST_CONSTEXPR_CXX17 constexpr
-#else
-# define TEST_CONSTEXPR_CXX17
-#endif
-
-#if TEST_STD_VER >= 20
@@ -187,13 +138 @@
-#if TEST_STD_VER >= 23
-#  define TEST_CONSTEXPR_CXX23 constexpr
-#else
-#  define TEST_CONSTEXPR_CXX23
-#endif
-
-#if TEST_STD_VER >= 26
-#  define TEST_CONSTEXPR_CXX26 constexpr
-#else
-#  define TEST_CONSTEXPR_CXX26
-#endif
-
-#define TEST_ALIGNAS_TYPE(...) TEST_ALIGNAS(TEST_ALIGNOF(__VA_ARGS__))
+#define TEST_ALIGNAS_TYPE(...) alignas(alignof(__VA_ARGS__))
@@ -217,2 +156,2 @@
-#if TEST_HAS_FEATURE(address_sanitizer) || TEST_HAS_FEATURE(hwaddress_sanitizer) || \
-    TEST_HAS_FEATURE(memory_sanitizer) || TEST_HAS_FEATURE(thread_sanitizer)
+#if TEST_HAS_FEATURE(address_sanitizer) || TEST_HAS_FEATURE(memory_sanitizer) || \
+    TEST_HAS_FEATURE(thread_sanitizer)
@@ -220,11 +159,12 @@
-#define TEST_IS_EXECUTED_IN_A_SLOW_ENVIRONMENT
-#endif
-
-#ifdef _LIBCPP_USE_FROZEN_CXX03_HEADERS
-#  ifdef _LIBCPP_HAS_NO_ALIGNED_ALLOCATION
-#    define TEST_HAS_NO_ALIGNED_ALLOCATION
-#  endif
-#elif defined(_LIBCPP_VERSION) && !_LIBCPP_HAS_ALIGNED_ALLOCATION
-#  define TEST_HAS_NO_ALIGNED_ALLOCATION
-#elif TEST_STD_VER < 17 && (!defined(__cpp_aligned_new) || __cpp_aligned_new < 201606L)
-#  define TEST_HAS_NO_ALIGNED_ALLOCATION
+#endif
+
+#if defined(_LIBCPP_NORETURN)
+#define TEST_NORETURN _LIBCPP_NORETURN
+#else
+#define TEST_NORETURN [[noreturn]]
+#endif
+
+#if defined(_LIBCPP_HAS_NO_ALIGNED_ALLOCATION) || \
+  (!(TEST_STD_VER > 14 || \
+    (defined(__cpp_aligned_new) && __cpp_aligned_new >= 201606L)))
+#define TEST_HAS_NO_ALIGNED_ALLOCATION
@@ -234,11 +174,12 @@
-#  define TEST_CONSTINIT constinit
-#elif __has_cpp_attribute(clang::require_constant_initialization)
-#  define TEST_CONSTINIT [[clang::require_constant_initialization]]
-#else
-#  define TEST_CONSTINIT
-#endif
-
-#if TEST_STD_VER < 11
-#define ASSERT_NOEXCEPT(...)
-#define ASSERT_NOT_NOEXCEPT(...)
-#else
+#define TEST_CONSTINIT constinit
+#elif defined(_LIBCPP_CONSTINIT)
+#define TEST_CONSTINIT _LIBCPP_CONSTINIT
+#else
+#define TEST_CONSTINIT
+#endif
+
+#if !defined(__cpp_impl_three_way_comparison) \
+    && (!defined(_MSC_VER) || defined(__clang__) || _MSC_VER < 1920 || _MSVC_LANG <= 201703L)
+#define TEST_HAS_NO_SPACESHIP_OPERATOR
+#endif
+
@@ -250 +190,0 @@
-#endif
@@ -260,17 +200,5 @@
-#define LIBCPP_ASSERT(...) static_assert(true, "")
-#define LIBCPP_STATIC_ASSERT(...) static_assert(true, "")
-#define LIBCPP_ASSERT_NOEXCEPT(...) static_assert(true, "")
-#define LIBCPP_ASSERT_NOT_NOEXCEPT(...) static_assert(true, "")
-#define LIBCPP_ONLY(...) static_assert(true, "")
-#endif
-
-#ifdef _LIBCPP_USE_FROZEN_CXX03_HEADERS
-#  define LIBCPP_NON_FROZEN_ASSERT(...) static_assert(true, "")
-#else
-#  define LIBCPP_NON_FROZEN_ASSERT(...) LIBCPP_ASSERT(__VA_ARGS__)
-#endif
-
-#if __has_cpp_attribute(nodiscard)
-#  define TEST_NODISCARD [[nodiscard]]
-#else
-#  define TEST_NODISCARD
+#define LIBCPP_ASSERT(...) static_assert(true)
+#define LIBCPP_STATIC_ASSERT(...) static_assert(true)
+#define LIBCPP_ASSERT_NOEXCEPT(...) static_assert(true)
+#define LIBCPP_ASSERT_NOT_NOEXCEPT(...) static_assert(true)
+#define LIBCPP_ONLY(...) static_assert(true)
@@ -280,6 +207,0 @@
-
-#ifdef _LIBCPP_USE_FROZEN_CXX03_HEADERS
-// from-chars is a C++17 feature, so it's never available anyways
-#elif !defined(_LIBCPP_VERSION) || _LIBCPP_AVAILABILITY_HAS_FROM_CHARS_FLOATING_POINT
-#  define TEST_HAS_FROM_CHARS_FLOATING_POINT
-#endif
@@ -310,6 +231,0 @@
-// This function can be used to hide some objects from compiler optimizations.
-//
-// For example, this is useful to hide the result of a call to `new` and ensure
-// that the compiler doesn't elide the call to new/delete. Otherwise, elliding
-// calls to new/delete is allowed by the Standard and compilers actually do it
-// when optimizations are enabled.
@@ -317,8 +233,3 @@
-inline Tp const& DoNotOptimize(Tp const& value) {
-  // The `m` constraint is invalid in the AMDGPU backend.
-#  if defined(__AMDGPU__) || defined(__NVPTX__)
-  asm volatile("" : : "r"(value) : "memory");
-#  else
-  asm volatile("" : : "r,m"(value) : "memory");
-#  endif
-  return value;
+inline
+void DoNotOptimize(Tp const& value) {
+    asm volatile("" : : "r,m"(value) : "memory");
@@ -328,7 +239,2 @@
-inline Tp& DoNotOptimize(Tp& value) {
-  // The `m` and `r` output constraint is invalid in the AMDGPU backend as well
-  // as i8 / i1 arguments, so we just capture the pointer instead.
-#  if defined(__AMDGPU__)
-  Tp* tmp = &value;
-  asm volatile("" : "+v"(tmp) : : "memory");
-#  elif defined(__clang__)
+inline void DoNotOptimize(Tp& value) {
+#if defined(__clang__)
@@ -336 +242 @@
-#  else
+#else
@@ -338,2 +244 @@
-#  endif
-  return value;
+#endif
@@ -344 +249 @@
-inline Tp const& DoNotOptimize(Tp const& value) {
+inline void DoNotOptimize(Tp const& value) {
@@ -348 +252,0 @@
-  return value;
@@ -412,6 +316,6 @@
-#if defined(_LIBCPP_VERSION) && !_LIBCPP_HAS_WIDE_CHARACTERS
-#  define TEST_HAS_NO_WIDE_CHARACTERS
-#endif
-
-#if defined(_LIBCPP_VERSION) && !_LIBCPP_HAS_UNICODE
-#  define TEST_HAS_NO_UNICODE
+#if defined(_LIBCPP_HAS_NO_WIDE_CHARACTERS)
+#   define TEST_HAS_NO_WIDE_CHARACTERS
+#endif
+
+#if defined(_LIBCPP_HAS_NO_UNICODE)
+#   define TEST_HAS_NO_UNICODE
@@ -419,20 +323,8 @@
-#  define TEST_HAS_NO_UNICODE
-#endif
-
-#ifdef _LIBCPP_USE_FROZEN_CXX03_HEADERS
-#  ifdef _LIBCPP_HAS_OPEN_WITH_WCHAR
-#    define TEST_HAS_OPEN_WITH_WCHAR
-#  endif
-#elif defined(_LIBCPP_VERSION) && _LIBCPP_HAS_OPEN_WITH_WCHAR
-#  define TEST_HAS_OPEN_WITH_WCHAR
-#endif
-
-#ifdef _LIBCPP_USE_FROZEN_CXX03_HEADERS
-#  ifdef _LIBCPP_HAS_NO_INT128
-#    define TEST_HAS_NO_INT128
-#  endif
-#elif (defined(_LIBCPP_VERSION) && !_LIBCPP_HAS_INT128) || defined(_MSVC_STL_VERSION)
-#  define TEST_HAS_NO_INT128
-#endif
-
-#if defined(_LIBCPP_VERSION) && !_LIBCPP_HAS_LOCALIZATION
+#   define TEST_HAS_NO_UNICODE
+#endif
+
+#if defined(_LIBCPP_HAS_NO_INT128) || defined(_MSVC_STL_VERSION)
+#   define TEST_HAS_NO_INT128
+#endif
+
+#if defined(_LIBCPP_HAS_NO_LOCALIZATION)
@@ -442 +334 @@
-#if TEST_STD_VER <= 17 || !defined(__cpp_char8_t)
+#if TEST_STD_VER <= 17 || !defined(__cpp_char8_t) || !defined(__cpp_lib_char8_t)
@@ -446 +338 @@
-#if defined(_LIBCPP_VERSION) && !_LIBCPP_HAS_THREADS
+#if defined(_LIBCPP_HAS_NO_THREADS)
@@ -450,25 +342,6 @@
-#if defined(_LIBCPP_VERSION) && !_LIBCPP_HAS_FILESYSTEM
-#  define TEST_HAS_NO_FILESYSTEM
-#endif
-
-#ifdef _LIBCPP_USE_FROZEN_CXX03_HEADERS
-#  ifdef _LIBCPP_HAS_NO_C8RTOMB_MBRTOC8
-#    define TEST_HAS_NO_C8RTOMB_MBRTOC8
-#  endif
-#elif defined(_LIBCPP_VERSION) && !_LIBCPP_HAS_C8RTOMB_MBRTOC8
-#  define TEST_HAS_NO_C8RTOMB_MBRTOC8
-#endif
-
-#if defined(_LIBCPP_VERSION) && !_LIBCPP_HAS_RANDOM_DEVICE
-#  define TEST_HAS_NO_RANDOM_DEVICE
-#endif
-
-#ifdef _LIBCPP_USE_FROZEN_CXX03_HEADERS
-// This is a C++20 feature, so it's never available anyways
-#  define TEST_HAS_NO_EXPERIMENTAL_TZDB
-#elif defined(_LIBCPP_VERSION) && !_LIBCPP_HAS_EXPERIMENTAL_TZDB
-#  define TEST_HAS_NO_EXPERIMENTAL_TZDB
-#endif
-
-#if defined(_LIBCPP_VERSION) && !_LIBCPP_HAS_TIME_ZONE_DATABASE
-#  define TEST_HAS_NO_TIME_ZONE_DATABASE
+#if defined(_LIBCPP_HAS_NO_FILESYSTEM_LIBRARY)
+#  define TEST_HAS_NO_FILESYSTEM_LIBRARY
+#endif
+
+#if defined(_LIBCPP_HAS_NO_FGETPOS_FSETPOS)
+#  define TEST_HAS_NO_FGETPOS_FSETPOS
@@ -511,40 +384,42 @@
-#ifdef _LIBCPP_SHORT_WCHAR
-#  define TEST_SHORT_WCHAR
-#endif
-
-#ifdef _LIBCPP_ABI_MICROSOFT
-#  define TEST_ABI_MICROSOFT
-#endif
-
-// This is a temporary workaround for user-defined `operator new` definitions
-// not being picked up on Apple platforms in some circumstances. This is under
-// investigation and should be short-lived.
-#ifdef __APPLE__
-#  define TEST_WORKAROUND_BUG_109234844_WEAK __attribute__((weak))
-#else
-#  define TEST_WORKAROUND_BUG_109234844_WEAK /* nothing */
-#endif
-
-#ifdef _AIX
-#  define TEST_IF_AIX(arg_true, arg_false) arg_true
-#else
-#  define TEST_IF_AIX(arg_true, arg_false) arg_false
-#endif
-
-// Placement `operator new`/`operator new[]` are not yet constexpr in C++26
-// when using MS ABI, because they are from <vcruntime_new.h>.
-#if defined(__cpp_lib_constexpr_new) && __cpp_lib_constexpr_new >= 202406L
-#  define TEST_CONSTEXPR_OPERATOR_NEW constexpr
-#else
-#  define TEST_CONSTEXPR_OPERATOR_NEW
-#endif
-
-#if defined(_MSC_VER) || __SIZEOF_LONG_DOUBLE__ == __SIZEOF_DOUBLE__
-#  define TEST_LONG_DOUBLE_IS_DOUBLE
-#endif
-
-#if defined(__LDBL_MANT_DIG__) && __LDBL_MANT_DIG__ == 64
-#  define TEST_LONG_DOUBLE_IS_80_BIT
-#endif
-
-#endif // SUPPORT_TEST_MACROS_HPP
+#define TEST_PREPARE_CALLABLE(std_algo_name) \
+    [](auto&&... __args) { return std_algo_name(std::forward<decltype(__args)>(__args)...); }
+
+#if TEST_STD_VER >= 14
+# define TEST_CONSTEXPR_CXX14 constexpr
+#else
+# define TEST_CONSTEXPR_CXX14
+#endif
+
+#if TEST_STD_VER >= 11
+# define TEST_ALIGNOF(...) alignof(__VA_ARGS__)
+# define TEST_ALIGNAS(...) alignas(__VA_ARGS__)
+# define TEST_CONSTEXPR constexpr
+# define TEST_NOEXCEPT noexcept
+# define TEST_NOEXCEPT_FALSE noexcept(false)
+# define TEST_NOEXCEPT_COND(...) noexcept(__VA_ARGS__)
+#else
+#   if defined(TEST_COMPILER_CLANG)
+#    define TEST_ALIGNOF(...) _Alignof(__VA_ARGS__)
+#   else
+#    define TEST_ALIGNOF(...) __alignof(__VA_ARGS__)
+#   endif
+# define TEST_ALIGNAS(...) __attribute__((__aligned__(__VA_ARGS__)))
+# define TEST_CONSTEXPR
+# define TEST_NOEXCEPT throw()
+# define TEST_NOEXCEPT_FALSE
+# define TEST_NOEXCEPT_COND(...)
+#endif
+
+#if TEST_STD_VER >= 17
+# define TEST_CONSTEXPR_CXX17 constexpr
+#else
+# define TEST_CONSTEXPR_CXX17
+#endif
+
+#if TEST_STD_VER >= 20
+# define TEST_CONSTEXPR_CXX20 constexpr
+#else
+# define TEST_CONSTEXPR_CXX20
+#endif
+
+#endif // _SUPPORT_TEST_MACROS_H