--- LLVM/libcxx/test/std/ranges/range.adaptors/range_adaptor_types.h
+++ oneDPL/test/parallel_api/ranges/range.zip/types.h
@@ -0,0 +1 @@
+// -*- C++ -*-
@@ -1,0 +3,7 @@
+//
+// Copyright (C) Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// This file incorporates work covered by the following copyright and permission
+// notice:
@@ -5 +12,0 @@
-// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
@@ -9,2 +16,2 @@
-#ifndef TEST_STD_RANGES_RANGE_ADAPTORS_RANGE_ADAPTOR_TYPES_H
-#define TEST_STD_RANGES_RANGE_ADAPTORS_RANGE_ADAPTOR_TYPES_H
+#ifndef TEST_STD_RANGES_RANGE_ADAPTORS_RANGE_ZIP_TYPES_H
+#define TEST_STD_RANGES_RANGE_ADAPTORS_RANGE_ZIP_TYPES_H
@@ -15,7 +22,3 @@
-#include "test_macros.h"
-#include "test_iterators.h"
-#include "test_range.h"
-
-#if TEST_STD_VER <= 20
-#  error "range.adaptor/types.h" can only be included in builds supporting C++20
-#endif // TEST_STD_VER <= 20
+#include "support/test_macros.h"
+#include "support/test_iterators.h"
+#include "support/test_range.h"
@@ -30,2 +32,0 @@
-
-  constexpr BufferView(T* b, std::size_t s) : buffer_(b), size_(s) {}
@@ -37 +38 @@
-struct Common : IntBufferView {
+struct Common :  IntBufferView {
@@ -53 +54 @@
-using SimpleCommon    = Common<true>;
+using SimpleCommon = Common<true>;
@@ -56 +57 @@
-using SimpleCommonRandomAccessSized    = SimpleCommon;
+using SimpleCommonRandomAccessSized = SimpleCommon;
@@ -69 +70 @@
-  using iterator       = forward_iterator<int*>;
+  using iterator = forward_iterator<int*>;
@@ -71,2 +72 @@
-    requires(!Simple)
-  {
+    requires(!Simple) {
@@ -77,2 +77 @@
-    requires(!Simple)
-  {
+    requires(!Simple) {
@@ -84 +83 @@
-using SimpleCommonNonRandom    = CommonNonRandom<true>;
+using SimpleCommonNonRandom = CommonNonRandom<true>;
@@ -97,2 +96 @@
-    requires(!Simple)
-  {
+    requires(!Simple) {
@@ -103,2 +101 @@
-    requires(!Simple)
-  {
+    requires(!Simple) {
@@ -110 +107 @@
-using SimpleNonCommon    = NonCommon<true>;
+using SimpleNonCommon = NonCommon<true>;
@@ -123,2 +120 @@
-    requires(!Simple)
-  {
+    requires(!Simple) {
@@ -129,2 +125 @@
-    requires(!Simple)
-  {
+    requires(!Simple) {
@@ -137 +132 @@
-using SimpleNonCommonSized                = NonCommonSized<true>;
+using SimpleNonCommonSized = NonCommonSized<true>;
@@ -139 +134 @@
-using NonSimpleNonCommonSized             = NonCommonSized<false>;
+using NonSimpleNonCommonSized = NonCommonSized<false>;
@@ -153 +148 @@
-  using iterator       = forward_iterator<int*>;
+  using iterator = forward_iterator<int*>;
@@ -156,2 +151 @@
-    requires(!Simple)
-  {
+    requires(!Simple) {
@@ -162,2 +156 @@
-    requires(!Simple)
-  {
+    requires(!Simple) {
@@ -171 +164 @@
-using SimpleNonCommonNonRandom    = NonCommonNonRandom<true>;
+using SimpleNonCommonNonRandom = NonCommonNonRandom<true>;
@@ -185,2 +178 @@
-    requires(!std::is_same_v<Iter, NonConstIter>)
-  {
+    requires(!std::is_same_v<Iter, NonConstIter>) {
@@ -192,2 +184 @@
-    requires(!std::is_same_v<Sent, NonConstSent>)
-  {
+    requires(!std::is_same_v<Sent, NonConstSent>) {
@@ -215,4 +206,4 @@
-  using value_type        = int;
-  using difference_type   = std::intptr_t;
-  using pointer           = Base;
-  using reference         = decltype(*Base{});
+  using value_type = int;
+  using difference_type = std::intptr_t;
+  using pointer = Base;
+  using reference = decltype(*Base{});
@@ -247,5 +238,2 @@
-using NonSimpleForwardSizedView =
-    BasicView<forward_sized_iterator<const int*>,
-              forward_sized_iterator<const int*>,
-              forward_sized_iterator<int*>,
-              forward_sized_iterator<int*>>;
+using NonSimpleForwardSizedView = BasicView<forward_sized_iterator<const int*>, forward_sized_iterator<const int*>,
+                                            forward_sized_iterator<int*>, forward_sized_iterator<int*>>;
@@ -266,4 +254,2 @@
-    BasicView<forward_sized_iterator<const int*>,
-              sized_sentinel<forward_sized_iterator<const int*>>,
-              forward_sized_iterator<int*>,
-              sized_sentinel<forward_sized_iterator<int*>>>;
+    BasicView<forward_sized_iterator<const int*>, sized_sentinel<forward_sized_iterator<const int*>>,
+              forward_sized_iterator<int*>, sized_sentinel<forward_sized_iterator<int*>>>;
@@ -298,4 +284,2 @@
-    BasicView<random_access_iterator<const int*>,
-              sentinel_wrapper<random_access_iterator<const int*>>,
-              random_access_iterator<int*>,
-              sentinel_wrapper<random_access_iterator<int*>> >;
+    BasicView<random_access_iterator<const int*>, sentinel_wrapper<random_access_iterator<const int*>>,
+              random_access_iterator<int*>, sentinel_wrapper<random_access_iterator<int*>> >;
@@ -330,5 +314,2 @@
-using NonSimpleInputCommonView =
-    BasicView<common_input_iterator<const int*>,
-              common_input_iterator<const int*>,
-              common_input_iterator<int*>,
-              common_input_iterator<int*>>;
+using NonSimpleInputCommonView = BasicView<common_input_iterator<const int*>, common_input_iterator<const int*>,
+                                           common_input_iterator<int*>, common_input_iterator<int*>>;
@@ -347,4 +328,2 @@
-    BasicView<common_input_iterator<const int*>,
-              sentinel_wrapper<common_input_iterator<const int*>>,
-              common_input_iterator<int*>,
-              sentinel_wrapper<common_input_iterator<int*>>>;
+    BasicView<common_input_iterator<const int*>, sentinel_wrapper<common_input_iterator<const int*>>,
+              common_input_iterator<int*>, sentinel_wrapper<common_input_iterator<int*>>>;
@@ -363,5 +342,2 @@
-using NonSimpleBidiCommonView =
-    BasicView<bidirectional_iterator<const int*>,
-              bidirectional_iterator<const int*>,
-              bidirectional_iterator<int*>,
-              bidirectional_iterator<int*>>;
+using NonSimpleBidiCommonView = BasicView<bidirectional_iterator<const int*>, bidirectional_iterator<const int*>,
+                                          bidirectional_iterator<int*>, bidirectional_iterator<int*>>;
@@ -402,4 +378,2 @@
-    BasicView<bidirectional_iterator<const int*>,
-              sentinel_wrapper<bidirectional_iterator<const int*>>,
-              bidirectional_iterator<int*>,
-              sentinel_wrapper<bidirectional_iterator<int*>>>;
+    BasicView<bidirectional_iterator<const int*>, sentinel_wrapper<bidirectional_iterator<const int*>>,
+              bidirectional_iterator<int*>, sentinel_wrapper<bidirectional_iterator<int*>>>;
@@ -420,4 +394,2 @@
-    BasicView<bidirectional_iterator<const int*>,
-              sized_sentinel<bidirectional_iterator<const int*>>,
-              bidirectional_iterator<int*>,
-              sized_sentinel<bidirectional_iterator<int*>>>;
+    BasicView<bidirectional_iterator<const int*>, sized_sentinel<bidirectional_iterator<const int*>>,
+              bidirectional_iterator<int*>, sized_sentinel<bidirectional_iterator<int*>>>;
@@ -430 +402 @@
-namespace adltest {
+namespace adltest{
@@ -432,11 +404,7 @@
-  int* iter_move_called_times = nullptr;
-  int* iter_swap_called_times = nullptr;
-  int i                       = 0;
-
-  constexpr iter_move_swap_iterator() = default;
-  constexpr iter_move_swap_iterator(int& iter_move_called_times_ref, int& iter_swap_called_times_ref)
-      : iter_move_called_times(&iter_move_called_times_ref),
-        iter_swap_called_times(&iter_swap_called_times_ref),
-        i(0) {}
-
-  using value_type      = int;
+
+  std::reference_wrapper<int> iter_move_called_times;
+  std::reference_wrapper<int> iter_swap_called_times;
+  int i = 0;
+
+  using iterator_category = std::input_iterator_tag;
+  using value_type = int;
@@ -451,5 +419 @@
-  constexpr iter_move_swap_iterator operator++(int) {
-    auto tmp = *this;
-    ++*this;
-    return tmp;
-  }
+  constexpr void operator++(int) { ++i; }
@@ -458,3 +421,0 @@
-  friend constexpr bool operator==(const iter_move_swap_iterator& x, const iter_move_swap_iterator& y) {
-    return x.i == y.i;
-  }
@@ -463 +424 @@
-    ++*it.iter_move_called_times;
+    ++it.iter_move_called_times;
@@ -467,2 +428,2 @@
-    ++*x.iter_swap_called_times;
-    ++*y.iter_swap_called_times;
+    ++x.iter_swap_called_times;
+    ++y.iter_swap_called_times;
@@ -480,78 +441 @@
-template <class Base = int*>
-struct convertible_forward_sized_iterator {
-  Base it_ = nullptr;
-
-  using iterator_category = std::forward_iterator_tag;
-  using value_type        = int;
-  using difference_type   = std::intptr_t;
-
-  convertible_forward_sized_iterator() = default;
-  constexpr convertible_forward_sized_iterator(Base it) : it_(it) {}
-
-  template <std::convertible_to<Base> U>
-  constexpr convertible_forward_sized_iterator(const convertible_forward_sized_iterator<U>& it) : it_(it.it_) {}
-
-  constexpr decltype(*Base{}) operator*() const { return *it_; }
-
-  constexpr convertible_forward_sized_iterator& operator++() {
-    ++it_;
-    return *this;
-  }
-  constexpr convertible_forward_sized_iterator operator++(int) { return forward_sized_iterator(it_++); }
-
-  friend constexpr bool
-  operator==(const convertible_forward_sized_iterator&, const convertible_forward_sized_iterator&) = default;
-
-  friend constexpr difference_type
-  operator-(const convertible_forward_sized_iterator& x, const convertible_forward_sized_iterator& y) {
-    return x.it_ - y.it_;
-  }
-};
-static_assert(std::forward_iterator<convertible_forward_sized_iterator<>>);
-
-template <class Base>
-struct convertible_sized_sentinel {
-  Base base_;
-  explicit convertible_sized_sentinel() = default;
-  constexpr convertible_sized_sentinel(const Base& it) : base_(it) {}
-
-  template <std::convertible_to<Base> U>
-  constexpr convertible_sized_sentinel(const convertible_sized_sentinel<U>& other) : base_(other.base_) {}
-
-  template <class U>
-    requires(std::convertible_to<Base, U> || std::convertible_to<U, Base>)
-  friend constexpr bool operator==(const convertible_sized_sentinel& s, const U& base) {
-    return s.base_ == base;
-  }
-  template <class U>
-    requires(std::convertible_to<Base, U> || std::convertible_to<U, Base>)
-  friend constexpr auto operator-(const convertible_sized_sentinel& s, const U& i) {
-    return s.base_ - i;
-  }
-
-  template <class U>
-    requires(std::convertible_to<Base, U> || std::convertible_to<U, Base>)
-  friend constexpr auto operator-(const U& i, const convertible_sized_sentinel& s) {
-    return i - s.base_;
-  }
-};
-static_assert(std::sized_sentinel_for<convertible_sized_sentinel<convertible_forward_sized_iterator<>>,
-                                      convertible_forward_sized_iterator<>>);
-static_assert(std::sized_sentinel_for<convertible_sized_sentinel<convertible_forward_sized_iterator<const int*>>,
-                                      convertible_forward_sized_iterator<int*>>);
-static_assert(std::sized_sentinel_for<convertible_sized_sentinel<convertible_forward_sized_iterator<int*>>,
-                                      convertible_forward_sized_iterator<const int*>>);
-
-struct ConstCompatibleForwardSized : IntBufferView {
-  using IntBufferView::IntBufferView;
-
-  using iterator       = convertible_forward_sized_iterator<int*>;
-  using const_iterator = convertible_forward_sized_iterator<const int*>;
-
-  constexpr iterator begin() { return {buffer_}; }
-  constexpr const_iterator begin() const { return {buffer_}; }
-  constexpr convertible_sized_sentinel<iterator> end() { return iterator{buffer_ + size_}; }
-  constexpr convertible_sized_sentinel<const_iterator> end() const { return const_iterator{buffer_ + size_}; }
-};
-
-#endif //  TEST_STD_RANGES_RANGE_ADAPTORS_RANGE_ADAPTOR_TYPES_H
+#endif // TEST_STD_RANGES_RANGE_ADAPTORS_RANGE_ZIP_TYPES_H